src/parser/view_transform.c

```diff
@@ -3162,6 +3207,68 @@ pt_copypush_terms (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE * query, PT_
   return;
 }
 
+/*
+ * mq_copypush_sargable_terms_dblink() -
+ *   return:
+ *   parser(in):
+ *   statement(in):
+ *   spec(in):
+ *   new_query(in/out):
+ *   infop(in):
+ */
+static int
+mq_copypush_sargable_terms_dblink (PARSER_CONTEXT * parser, PT_NODE * statement, PT_NODE * spec, PT_NODE * new_query,
+				   FIND_ID_INFO * infop)
+{
+  PT_NODE *term, *new_term, *push_term_list;
+  int push_cnt, copy_cnt;
+  PT_NODE *save_next;
+
+  /* init */
+  push_term_list = NULL;
+  push_cnt = 0;
+  copy_cnt = -1;
+
+  for (term = statement->info.query.q.select.where; term; term = term->next)
+    {
+      if (pt_sargable_term (parser, term, infop) && PT_PUSHABLE_TERM (infop))
+	{
+	  /* copy term */
+	  new_term = parser_copy_tree (parser, term);
+
+	  /* for term, mark as copy-pushed term */
+	  if (term->node_type == PT_EXPR)
+	    {
+	      if (copy_cnt == -1)	/* very the first time */
+		{
+		  copy_cnt = pt_check_copypush_subquery (parser, new_query);
+		}
+
+	      if (copy_cnt == 0)	/* not found not-pushable query */
+		{
+		  PT_EXPR_INFO_SET_FLAG (term, PT_EXPR_INFO_COPYPUSH);
```

Comment:
There is nowhere to use PT_EXPR_INFO_COPYPUSH.  it is cleared immediately after setting it.

Reply:
PT_EXPR_INFO_SET_FLAG  is used after push down done for removing the term node (which is already pushed to the dblink).

Reply:
Should copy_cnt be checked in dblink?

Reply:
the copy_cnt is used to set the flag PT_EXPR_INFO_COPYPUSH, the flag is used to check the term is pushed down and to remove the node.

Reply:
The call of pt_check_copypush_subquery () is not needed in dblink. The copy_cnt is always 0.

Reply:
copy_cnt is always 0 in dblink, so the below code looks better.

	  /* for term, mark as copy-pushed term */
	  if (term->node_type == PT_EXPR)
	    {
	      PT_EXPR_INFO_SET_FLAG (term, PT_EXPR_INFO_COPYPUSH);
	      PT_EXPR_INFO_CLEAR_FLAG (new_term, PT_EXPR_INFO_COPYPUSH);
	    }

Reply:
'PT_EXPR_INFO_CLEAR_FLAG (new_term, PT_EXPR_INFO_COPYPUSH);' is not needed in dblink.

In my opinion it is better not to delete predicates in dblink. This helps when predicate push is not working properly. This is my opinion, the decision is yours.

Reply:
Ok, I agree you.